### ISSUE TITLE
waiting for stream to close.

### ISSUE THREAD

**Issue by jimmywarting** [MattiasBuelens/remote-web-streams#27]
it seems like as soon as all chunks have been transfered to the other remote streams it closes without actually knowing if the worker is done or not?

i get race condition errors...

```js
await redable.pipeTo(writable) // resolves earlier than it should...?
// here i'm required to wait a bit longer for the worker to successfully write it to indexeddb.
await new Promise(rs => setTimeout(rs, ms))
file = await idb.get(key)
```

maybe it's a bug or just not implemented yet?

**Comment by MattiasBuelens** [comment]
I'm assuming `writable` is a `RemoteWritableStream`'s writable end, whose readable end is being read inside a worker that pushes stuff into IndexedDB?

Unfortunately, there's not a whole lot I can do about this. `RemoteWritableStream`'s sink `close()` method [sends a message to the other end](https://github.com/MattiasBuelens/remote-web-streams/blob/v0.2.0/src/writable.ts#L55-L61) to call `ReadableStreamDefaultController.close()`, but that doesn't say anything about *when* that closing is observed. For example, there may be multiple chunks in the readable stream's queue that have not yet been consumed.

Note that this is not specific to remote streams. When you make a pipe chain of "regular" streams, the first `pipeTo()` promise will resolve *before* the subsequent `pipeTo()` promises, without waiting for the "final sink" to process the last chunk. That's why you usually only await the *last* `pipeTo()` promise, as in:
```javascript
await readable.pipeThrough(transform1).pipeThrough(transform2).pipeTo(writable);
```
(There's some discussion in whatwg/streams#976 about possibly adding a `finally()` method to an underlying source, which might help.)

For now, your best best would be to have the worker send a message back to the main thread when it's *actually* done writing all chunks to IndexedDB. Perhaps you can (ab)use another remote stream for that? 😛

**Comment by jimmywarting** [comment]
Hmm, roger that. 
gone close this as there seems to be nothing to do about it right now...
maybe have to wait for that `finally()` method to arrive... or better yet for safari to start shipping transferable streams... only reason why this pkg is still needed. all other browser support it already.

For now i have rolled my own solution that sends a postMessage back notifying me when it's done.

>  this is not specific to remote streams. When you make a pipe chain of "regular" streams, the first pipeTo() promise will resolve before the subsequent pipeTo() promises, without waiting for the "final sink"

I figured it was something like that... i only really use one writable stream that dose not go trough any transformers.
`await redable.pipeTo(writable)` was just a very basic simplification of my problem. but the same thing still applies. 

I have basically built a polyfill for `FileSystemWritableFileStream` which extends WritableStream... so it not a transform stream with both a readable + writable combo. so sending all the data to the worker required me to write something like this:

```js
const worker = new Worker('./worker.js', { type: 'module' })
const remote = new RemoteWritableStream()
const writable = new FileSystemWritableFileStream(
  remote.writable.getWriter() // non normal option (just private / internal use)
)
postMessage('', [remote.readablePort])
return writable
```
As you can see i needed it to be a different kind of writable stream that do also have additional methods added to the public facing methods that also has a `seek / truncate / write / close` method added to it.

constructing a `RemoteWritableStream` creates one additional transformer. so now i can no longer safely wait for the closed promise to complete. 

So what i really need to be able to accomplish is more something in the lines of:



```js
// main.js
import { RemoteWritableStreamController } from 'remote-web-streams'
// this controller would basically refer to the controller constructed inside the worker.js (below)
const controller = new RemoteWritableStreamController()
worker.postMessage('', [controller.port])
return new FileSystemWritableFileStream(controller)

// worker.js
import { RemoteWritableStreamController } from 'remote-web-streams'
onmessage = evt => {
  RemoteWritableStreamController.from({
    async start() { },
    async close() { },
    async abort() { },
    async write(chunk, ctrl) { ... },
    port: evt.ports[0]
  })
}
```
So basically you would not create any additional writable/readable stream in the main thread. the controller will basically be built in the worker instead. 
i guess this would more likely resemble how transferring a stream to a worker would work by the browser without creating any additional pipes.

---

It may look a bit weird/backward to create the `FileSystemWritableFileStream` in the main thread and not constructing the `FileSystemWritableFileStream` in the worker and transferring it to the main thread. but we also have to remember that they are not transferable as well. (specially when they are polyfilled)

**Comment by jimmywarting** [comment]
if you are interested in my work then here is what i have been doing: https://github.com/jimmywarting/native-file-system-adapter/pull/62 (i did not end up using your package after all... it broke too many test)

**Comment by KingSora** [comment]
Good day! :)

The writable stream api is mostly async, so all the functions like `write`, `close` and `abort` returns promises.
Wouldn't a more accurate implementation look something like this:

```ts
export class MessagePortSink<W> implements UnderlyingSink<W> {
  constructor(private readonly _port: MessagePort) {}
  
  write(chunk: W, controller: WritableStreamDefaultController) {
    return new Promise((resolve, reject) => {
      this._port.addEventListener('message', ({ data }) => {
        if (data === /* message which indicates that data has been written */) {
          resolve();
        }
        if (data === /* message which indicates that data writing wasn't successful */) {
          reject();
        }
      });
      this._port.postMessage(/* message which indicates to write data */);
    });
  }
  //...
}
```

The same concept would apply to `close` and `abort`, instead of assuming the operation was instantly successful, a promise should be created which resolves / rejects in the point in the where the operation actually finished.
